src/polysolve/nonlinear/Criteria.hpp

```diff
@@ -0,0 +1,133 @@
+#pragma once
+
+#include <cstddef>
+#include <iostream>
+
+namespace polysolve::nonlinear
+{
+    // Source: https://github.com/PatWie/CppNumericalSolvers/blob/7eddf28fa5a8872a956d3c8666055cac2f5a535d/include/cppoptlib/meta.h
+    // License: MIT
+
+    enum class Status
+    {
+        NotStarted = -1,
+        Continue = 0,
+        IterationLimit,
+        XDeltaTolerance,
+        FDeltaTolerance,
+        GradNormTolerance,
+        Condition,
+        UserDefined
+    };
+
+    template <typename T>
+    class Criteria
+    {
+    public:
+        size_t iterations; //!< Maximum number of iterations
+        T xDelta;          //!< Minimum change in parameter vector
+        T fDelta;          //!< Minimum change in cost function
+        T gradNorm;        //!< Minimum norm of gradient vector
+        T condition;       //!< Maximum condition number of Hessian
+
+        Criteria()
+        {
+            reset();
+        }
+
+        static Criteria defaults()
```

Comment:
this is not necessary as we have the jse. we should have an invalid criteria and assert before using it